datadog.go

```diff
@@ -0,0 +1,237 @@
+	"go.opencensus.io/stats/view"
+	"go.opencensus.io/tag"
+)
+
+// Exporter exports stats to Datadog
+type Exporter struct {
+	opts      Options
+	collector *collector
+	client    *statsd.Client
+}
+
+// Options contains options for configuring the exporter
+type Options struct {
+	// Namespace to prepend to all metrics
+	Namespace string
+
+	// Host for DogStatsD connection
+	Host string
+
+	// Port for DogStatsD connection
+	Port string
+
+	// OnError is the hook to be called when there is
+	// an error occurred when uploading the stats data.
+	// If no custom hook is set, errors are logged.
+	// Optional.
+	OnError func(err error)
+
+	// Tags are global tags added to each metric
+	Tags []string
+}
+
+var (
+	newExporterOnce      sync.Once
+	errSingletonExporter = errors.New("expecting only one exporter per instance")
+)
+
+// NewExporter returns an exporter that exports stats to Datadog
+func NewExporter(o Options) (*Exporter, error) {
+	var err = errSingletonExporter
+	var exporter *Exporter
+	newExporterOnce.Do(func() {
+		exporter, err = newExporter(o)
+	})
+	return exporter, err
+}
+
+func getEndpoint(o Options) string {
+	var endpoint string
+	if o.Host != "" && o.Port != "" {
+		endpoint = o.Host + ":" + o.Port
+	} else {
+		endpoint = "127.0.0.1:8125"
+	}
+
+	log.Printf("Endpoint set at: %v", endpoint)
+	return endpoint
+}
+
+func newExporter(o Options) (*Exporter, error) {
+
+	endpoint := getEndpoint(o)
+
+	client, err := statsd.New(endpoint)
+	if err != nil {
+		log.Fatal(err)
+	}
+	collector := newCollector(o)
+
+	e := &Exporter{
+		opts:      o,
+		collector: collector,
+		client:    client,
+	}
+	return e, nil
+}
+
+// client implements datadog.Client
+type collector struct {
+	opts Options
+
+	// mu guards all the fields.
+	mu sync.Mutex
+
+	skipErrors bool
+
+	// viewData is accumulated and appended on every Export
+	// invocation from stats.
+	viewData map[string]*view.Data
+
+	viewsMu sync.Mutex
+
+	registeredViews map[string]string
+}
+
+func newCollector(o Options) *collector {
+	return &collector{
+		opts:            o,
+		registeredViews: make(map[string]string),
+		viewData:        make(map[string]*view.Data),
+	}
+}
+
+// ExportView exports to Datadog if view data has one or more rows.
+func (e *Exporter) ExportView(vd *view.Data) {
+	if len(vd.Rows) == 0 {
+		return
+	}
+	e.collector.addViewData(vd, e.client)
+}
+
+func (c *collector) registerViews(views ...*view.View) {
+	count := 0
+	for _, view := range views {
+
+		sig := viewSignature(c.opts.Namespace, view)
+		c.viewsMu.Lock()
+		_, ok := c.registeredViews[sig]
+		c.viewsMu.Unlock()
+
+		if !ok {
+			metadata := view.Description
+			c.viewsMu.Lock()
+			c.registeredViews[sig] = metadata
+			c.viewsMu.Unlock()
+			count++
+		}
+	}
+	if count == 0 {
+		return
+	}
+}
+
+func viewName(namespace string, v *view.View) string {
+	var name string
+	if namespace != "" {
+		name = namespace + "."
+	}
+	//return name + internal.Sanitize(v.Name)
```

Comment:
why is this commented?

Reply:
The `internal` package from the opencensus-go client has a `Sanitize()` function that replaces non-alphanumeric characters to underscores & truncates the string to 100 characters. I had to comment it out when we moved to our own repo.  Should we sanitize the metric name ourselves?

Reply:
Maybe I'm missing something but why can't we import and use it?

Reply:
I think it's because the import is an `internal package` of opencensus-go and since we're outside the parent of the `internal` directory we can't use it. I believe it's a design rule in go.

Reply:
TIL this works like this since 1.5 🙈